text/0072-kafka-schema-registry.md

```diff
@@ -0,0 +1,111 @@
+- Start Date: 2023-02-01
+- RFC Type: feature
+- RFC PR: n/a
+- RFC Status: draft
+
+# Summary
+
+This RFC proposes introducing a centralized schema repository for the Kafka topics in use at Sentry. It provides mappings between the Kafka topics we are using to the encoding format and schemas for messages on that topic.
+
+Topic registration and schema validation of messages will be optional for the foreseeable future.
+
+# Motivation
+
+Kafka is increasingly used as the message bus between services at Sentry. Kafka topics and their schemas are usually not internal to any one service, but part of the contract between services.
+
+As Sentry grows, the number of message types and topics in use is exploding. We are marching further towards increasingly distributed ownership of data by product and infrastructure engineering teams. We have already seen an increase in the number of incidents related with invalid data and schema issues at Sentry. It is reasonable to expect this trend to continue as the number of topics, data types, teams and engineers contributing to Sentry grows.
+
+The goals of the centralized schema repository are to:
+
+- provide an explicit, single source of truth about how data in any given topic should look, and the contract that all consumers and producers of a topic must adhere to
+- provide more stability by enforcing backward compatibility of changes via automation
+- explicitly enumerate ownership of schemas
+- make it easier for consumers to identify and reject bad messages without pausing the whole pipeline
+
+# Supporting Data
+
+Over the last quarter, we have seen many incidents related to schema disagreements at Sentry. Examples include the following: (note: inc numbers are internal to Sentry)
+
+- Post process forwarder (INC-218)
+- Snuba’s transactions consumer (INC-210)
+- Super big consumers (INC-220)
+- Replays consumer (INC-250, INC-281)
+
+Many of these incidents are P1. Since messages are in order, an invalid message often halts consumers and requires manual intervention. This is disruptive to both Sentry engineers and our users (it takes us much longer to recover).
+
+# Options Considered
+
+### **Option A (preferred): Publish a library**
+
+Schemas will be made available as a library for multiple programming languages (Python and Rust to start).
+
+**Client usage example:**
+
+```python
+from sentry_kafka_schemas import get_schema
+
+topic = "events"
+schema = get_schema(topic)
```

Comment:
Wanted to add something which might not be relevant on which option to choose. But more around how option A should look like. IMO, we shouldn't allow getting a schema without a version. Most folks won't know what this would mean. Is it the first version, or is it the latest version? If its the latest version, how can one know whether updating the schema library version causes backward incompatible change. Forcing versions in the api makes it explicit which version of schema you use. It also makes it easier finding consumers who might be on older versions more easily.

Reply:
I have been thinking about this quite a bit. I prefer to omit by default- though this is not a strong preference and I agree there are some benefits to being explicit. My thinking around this was:
- Looking at our usage of Kafka so far, I think at Sentry it's actually likely that we will never bump a version number *ever* on any topic. Though I added this example to this RFC, I don't actually expect our usage to look like this (unless we completely switch out the message format or something which would be a larger undertaking). I think it's more likely that we will continue to evolve our schema in a backward compatible way and none of the versions will ever go past 1
- Given that we never update the schema in a non backward compatible way, none of our messages actually contain version numbers. I would like to keep it this way rather than start forcing everyone to put extra metadata inside their messages.
- The alternative to putting the version in the message could be hardcoding the version number inside all of our codebases next to every topic. But I think this is a bit of extra work if we're never really bumping major versions anyway.
- The other option I considered was to remove `version` out of here entirely so that only backwards compatible changes are allowed. However, I think it's might be better to future proof even if it's not that likely we'll use it - it's still there if we want it later and we actually need to make a breaking change.

Reply:
If we don't have versioned payloads, would we consider having versioned topics instead? For example, when a producer needs to make a breaking change to a message format, the producer starts publishing to a new topic. This makes breaking changes more expensive but I think that's a benefit.

Reply:
The payload is still supposed to be versioned (just fixed the example to actually include this field) and the user can always request a specific version of the schema. Here we are just debating whether the schema version should default to latest if none is passed.

About the decision whether to publish an entire new topic or bump the major version instead - I think that is up to the developer to decide. The current design of the registry would allow for both options and doesn't really take an opinion about it either way.

Reply:
A case for storing the version on each message (possibly as a header - headers are for metadata, this seems a good use), applying a version to every schema change no matter on the compatibility model we choose and making the consumer declare the highest version they support. 

This approach would allow us to log warning and collect metrics about schema mismatches (even for backwards compatible changes). 
I define a schema mismatch as a consumer dealing with a message that has a higher version than the highest version it supports, even if the changes are all backwards compatible and the consumer can still deal with the data it receives.

I believe we are interested in highlighting these changes because they can hide semantic incompatibilities. For instance, let's imagine I introduce an data category on outcomes to separate dynamic sampling data from others, a consumer may ignore the new field and, for what concerns schema validation all would be good.
The application may be broken in a subtle way.

If, instead, I start seeing warning about a schema mismatch after adding this new field and forgetting to adapt the consumer business logic to take it into account, I have a chance to react and fix the application.

Reply:
I think this is a fair concern, but I think it may be a bit heavy handed for size of the problem today. Some arguments against creating new version for every schema change and putting the versions inside the messages:

- This behaviour (libraries out of sync in producer and consumer) is expected to happen under normal circumstances. It is impossible that all producers and consumers get rolled out in prod at the same moment, so any logging or metrics here will be incredibly noisy. It can even be normal for one to be updated even days ahead of another as the unfortunate reality at Sentry is that approval / deployment processes do end up taking a while. This is not really an issue as long as the changes are done in a backward compatible order.
- The problem of updating the library scales with the number of services to be updated. Today there is not too many. While it is possible in the future, I think this is a distant problem and not one we have today.
- I am concerned this repository will become too polluted with copies of almost identical schemas. There will be more friction involved in adding and changing schemas, and they will be harder to review. The library will grow too big.
- I would prefer to avoid enforcing changes to consumers/producers/message format in order for clients to adopt this repository. It will be harder to migrate onto the schema repository.
- I think bumping the version for every schema change adds complexity and lots of potential for edge cases in how the versioning system works. There would need to be more of a distinction between backwards compatible and non backwards compatible changes. We may need major and minor versions.

How about the following:
- In the short term, ensure we have an aggressive dependabot configuration set up on all services that use the schema repository. Pull request should be opened soon after the new version is available.
- If the number of services at Sentry grows, we can add the library version (or some other checksum) into the header to identify differences between producer and consumer versions without changing the whole schema versioning system.

Reply:
> In the short term, ensure we have an aggressive dependabot configuration set up on all services that use the schema repository. Pull request should be opened soon after the new version is available.

I'd be ok with this if we confirm that every owner of a schema can cut a release on their own without external approval.  Dependabot requires us to cut releases whenever we make changes to the schema.
If the process it annoying people will start depending on github dependency on specific commits in their requirements file instead of importing the package from pypi. At that point dependabot will be useless.

Reply:
Mind writing the outcome of this conversation: concern and mitigation, on the rfc itself. Otherwise the reasoning will be lost when it is approved.